### PR TITLE
Pull request 02102023

### PR DESCRIPTION
**Pull request by AnhQuoc2906**

### CODE REVIEW

File: Api/Controllers/RestAPI/UserController.cs [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+        public UserController(IUserServices userServices, IGenericRepository<Client> clientRepo, IGenericRepository<UserEntity> userRepo)
+        {
+            _userServices = userServices;
+            _clientRepo = clientRepo;
+            _userRepo = userRepo;
+        }
+        // GET: api/<UserController>
+        [Route("Client")]
+        [HttpGet]
+        public IEnumerable<string> Get()
```
Comment (danielNg91) [inline Api/Controllers/RestAPI/UserController.cs:26]: asp.net have problem when 2 controller have methods that the same name, Give methods in this class a more specific name

File: Api/Controllers/RestAPI/UserController.cs [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+        private readonly IGenericRepository<Client> _clientRepo;
+        private readonly IGenericRepository<UserEntity> _userRepo;
+        public UserController(IUserServices userServices, IGenericRepository<Client> clientRepo, IGenericRepository<UserEntity> userRepo)
+        {
+            _userServices = userServices;
+            _clientRepo = clientRepo;
+            _userRepo = userRepo;
+        }
+        // GET: api/<UserController>
+        [Route("Client")]
```
Comment (danielNg91) [inline Api/Controllers/RestAPI/UserController.cs:24]: remove these clients api

File: Api/Controllers/RestAPI/UserController.cs [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+            var users = await _userServices.GetUsers();
+            return Ok(users);
+        }
+
+        // GET api/User/5
+        [Route("User/{id}")]
+        [HttpGet]
+        public async Task<IActionResult> GetUser(string id)
+        {
+            try
```
Comment (danielNg91) [inline Api/Controllers/RestAPI/UserController.cs:70]: remove this try catch when you just use it for error handling

File: Application/Interfaces/ICharacterServices.cs [hunk 4]
```diff
@@ -0,0 +1,39 @@
+﻿using Domain.Entities.Character;
+
+namespace Application.Interfaces;
+
+public interface ICharacterServices
+{
+    //Character Asset
+    Task<ICollection<CharacterAsset>> GetCharacterAssets();
+    Task<CharacterAsset> GetCharacterAsset(Guid characterAssetId);
```
Comment (danielNg91) [inline Application/Interfaces/ICharacterServices.cs:9]: fix same as above

File: Application/Interfaces/IGameServices.cs [hunk 5]
```diff
@@ -0,0 +1,21 @@
+﻿using Domain.Entities.Game;
+
+namespace Application.Interfaces; 
+public interface IGameServices {
+    //Game
+    Task<ICollection<GameEntity>> GetGames();
```
Comment (danielNg91) [inline Application/Interfaces/IGameServices.cs:6]: fix same as above

File: Application/Interfaces/ITransactionServices.cs [hunk 6]
```diff
@@ -0,0 +1,14 @@
+﻿using Domain.Entities.Payment;
+using Domain.Entities.Transaction;
+
+namespace Application.Interfaces;
+public interface ITransactionServices {
+    //Transaction
+    Task<ICollection<TransactionEntity>> GetTransactions();
```
Comment (danielNg91) [inline Application/Interfaces/ITransactionServices.cs:7]: fix same as above

File: Application/Interfaces/IUserServices.cs [hunk 7]
```diff
@@ -0,0 +1,21 @@
+﻿using Domain.Entities.User;
+
+namespace Application.Interfaces; 
+public interface IUserServices {
+    //Clients
+    Task<ICollection<Client>> GetClients();
```
Comment (danielNg91) [inline Application/Interfaces/IUserServices.cs:6]: fix same as above, remove services for client

File: Application/Services/ActivityServices/ActivityServices.cs [hunk 8]
```diff
@@ -0,0 +1,137 @@
...
+{
+    public readonly IGenericRepository<ActivityEntity> _activityRepo;
+    public readonly IGenericRepository<ActivityType> _activityTypeRepo;
+    
+    public ActivityServices(IGenericRepository<ActivityEntity> activityRepo, IGenericRepository<ActivityType> activityTypeRepo)
+    {
+        _activityRepo = activityRepo;
+        _activityTypeRepo = activityTypeRepo;
+    }
+    public async Task<ICollection<ActivityEntity>> GetActivities() {
```
Comment (danielNg91) [inline Application/Services/ActivityServices/ActivityServices.cs:18]: this one line function is not necessary

File: Application/Services/ActivityServices/ActivityServices.cs [hunk 9]
```diff
@@ -0,0 +1,137 @@
...
+            throw new Exception($"Activity or activity type not found");
+        }
+        else
+        {
+            return activities;
+        }
+    }
+    public async Task<int> CountActivities()
+    {
+        var activities = await _activityRepo.ListAsync();
```
Comment (danielNg91) [inline Application/Services/ActivityServices/ActivityServices.cs:63]: useless oneline function, remove it otherswhere

File: Application/Services/ActivityServices/ActivityServices.cs [hunk 10]
```diff
@@ -0,0 +1,137 @@
...
+        if (activityType == null)
+        {
+            throw new Exception($"Activity type not found");
+        }
+        else
+        {
+            return activityType;
+        }
+    }
+    public async Task<ICollection<ActivityType>> GetActivityTypes(Guid id, int typeId)
```
Comment (danielNg91) [inline Application/Services/ActivityServices/ActivityServices.cs:97]: What is this function used for?

File: Application/Services/CharacterServices/CharacterServices.cs [hunk 11]
```diff
@@ -0,0 +1,248 @@
...
+            throw new Exception($"Character Asset or Asset Id / CharacterId not found");
+        }
+        else
+        {
+            return charAss;
+        }
+    }
+    public async Task<int> CountCharacterAssets()
+    {
+        var charAss = await _characterAssetRepo.ListAsync();
```
Comment (danielNg91) [inline Application/Services/CharacterServices/CharacterServices.cs:66]: we already have SumAsync function in Repository, why we need this implementation?

**Review by danielNg91 — CHANGES_REQUESTED** [review]

File: Domain/Entities/Activity/ActivityEntity.cs [hunk 12]
```diff
@@ -0,0 +1,22 @@
...
+namespace Domain.Entities.Activity;
+
+[Table("Activity")]
+public class ActivityEntity : BaseEntity
+{
+    public string Name { get; set; }
+    public int Status { get; set; }
+
+    // 1 Activity Type - M Activity
+
```
Comment (danielNg91) [inline Domain/Entities/Activity/ActivityEntity.cs:14]: remove this line break

File: Infrastructure/Contexts/EntityTypeConfigurations/EntityTypeConfiguration.cs [hunk 13]
```diff
@@ -0,0 +1,175 @@
...
+using Domain.Entities.Level;
+using Domain.Entities.Payment;
+using Domain.Entities.Transaction;
+using Domain.Entities.User;
+using Domain.Entities.Wallet;
+using Microsoft.EntityFrameworkCore;
+using Microsoft.EntityFrameworkCore.Metadata.Builders;
+
+namespace Infrastructure.Contexts;
+public class EntityTypeConfiguration :
```
Comment (danielNg91) [inline Infrastructure/Contexts/EntityTypeConfigurations/EntityTypeConfiguration.cs:15]: split to smaller files, just 1 class for 1 EntityTypeConfiguration (if needed)

**Review by danielNg91 — CHANGES_REQUESTED** [review]

File: Api/Controllers/RestAPI/ActivityController.cs [hunk 14]
```diff
@@ -0,0 +1,63 @@
...
+        public async Task<IActionResult> GetActivitíes()
+        {
+            var activities = await _activityServices.GetActivities();
+            if(activities.Any())
+            {
+                return Ok(activities);
+            }
+            else
+            {
+                throw new Exception("0 item in list");
```
[the remark this answers is not in the document]
Comment (AnhQuoc2906) [reply Api/Controllers/RestAPI/ActivityController.cs:34]: Remove if-else, return collection:

```
public async Task<IActionResult> GetActivitíes()
        {
            var activities = await _activityServices.GetActivities();
            return Ok(activities);
        }
```

File: Api/Controllers/RestAPI/AttributeController.cs [hunk 15]
```diff
@@ -0,0 +1,97 @@
...
+        public async Task<IActionResult> Get()
+        {
+            var attribute = await _attributeServices.GetAttributeGroups();
+            if (attribute.Any())
+            {
+                return Ok(attribute);
+            }
+            else
+            {
+                throw new Exception("0 item in list");
```
[the remark this answers is not in the document]
Comment (AnhQuoc2906) [reply Api/Controllers/RestAPI/AttributeController.cs:33]: Remove if-else, return collection:

```
public async Task<IActionResult> Get()
        {
            var attribute = await _attributeServices.GetAttributeGroups();
            return Ok(attribute);
        }
```

File: Api/Controllers/RestAPI/ActivityController.cs (hunk 14 above)
[the remark this answers is not in the document]
Comment (danielNg91) [reply Api/Controllers/RestAPI/ActivityController.cs:34]: yup, let return collection

File: Api/Controllers/RestAPI/AttributeController.cs [hunk 16]
```diff
@@ -0,0 +1,97 @@
...
+        }
+
+        // POST api/<AttributeController>
+        [HttpPost]
+        public async Task<IActionResult> Post([FromBody] AttributeGroup attributeGroup)
+        {
+            try
+            {
+                attributeGroup.Id = Guid.NewGuid();
+                attributeGroup.isDeleted = false;
```
[the remark this answers is not in the document]
Comment (AnhQuoc2906) [reply Api/Controllers/RestAPI/AttributeController.cs:67]: Set abstract class to auto-generate like this:
```
public abstract class BaseEntity
    {
        public Guid Id { get; set; } = Guid.NewGuid();
        public bool isDeleted { get; set; } = false; //false: not deleted, true: deleted
    }
```

And change the post method to this: 

```
var attGrps = _attributeServices.GetAttributeGroups();
                foreach(var attgrp in attGrps.Result)
                {
                    if(attgrp.Name == attributeGroup.Name)
                    {
                        return Conflict(attgrp);
                    } else
                    continue;
                }
                var attGrp = new AttributeGroup
                {
                    Name = attributeGroup.Name,
                    Effect = attributeGroup.Effect
                };
                await _attributeServices.CreateAttributeGroup(attGrp);
                return NoContent();
    
    
```

File: Api/Program.cs [hunk 17]
```diff
@@ -1,11 +1,27 @@
...
-
 builder.Services.AddControllers();
 // Learn more about configuring Swagger/OpenAPI at https://aka.ms/aspnetcore/swashbuckle
 builder.Services.AddEndpointsApiExplorer();
 builder.Services.AddSwaggerGen();
+builder.Services.AddDbContext<ApplicationDbContext>
+    (options => options.UseSqlServer("name=ConnectionStrings:GHub_Connection", b=>b.MigrationsAssembly("Infrastructure")));
+
+//-------------------------------
+//Add configurations to Services
```
[the remark this answers is not in the document]
Comment (AnhQuoc2906) [reply Api/Program.cs:19]: Change from comments to region like this
```
#region Add configurations to Services
var services = builder.Services;
services.AddAppServices();
#endregion
```

File: Domain/Entities/Activity/ActivityEntity.cs (hunk 12 above)
Comment (AnhQuoc2906) [reply Domain/Entities/Activity/ActivityEntity.cs:14]: Fixed: 

```
public int Status { get; set; }
    // 1 Activity Type - M Activity
    public Guid ActivityTypeId { get; set; }
```

File: Api/Configurations/AppConfigurationService.cs [hunk 18]
```diff
@@ -0,0 +1,44 @@
...
+using Application.Services.WalletServices;
+using Infrastructure.Repositories;
+
+namespace Api.Configurations
+{
+    public static class AppConfigurationService
+    {
+        public static void AddAppServices(this IServiceCollection services)
+        {
+            services.AddScoped(typeof(IGenericRepository<>), typeof(GenericRepository<>));
```
Comment (danielNg91) [inline Api/Configurations/AppConfigurationService.cs:21]: gud cho'p

File: Api/Controllers/ActivityController.cs [hunk 19]
```diff
@@ -0,0 +1,47 @@
+﻿using Application.Interfaces;
+using Application.Interfaces.Activity;
+using Domain.Entities.Activity;
+using Microsoft.AspNetCore.Authorization;
+using Microsoft.AspNetCore.Mvc;
+
+namespace Api.Controllers
+{
+    [AllowAnonymous]
```
Comment (danielNg91) [inline Api/Controllers/ActivityController.cs:9]: remove it, we don't use auths in this services

File: Api/Controllers/ActivityController.cs [hunk 20]
```diff
@@ -0,0 +1,47 @@
...
+        }
+        [HttpGet]
+        public async Task<IActionResult> GetActivitíes()
+        {
+            var activities = await _activityServices.List();
+            return Ok(activities);
+        }
+
+        [HttpGet("{id}")]
+        public string Get(int id)
```
Comment (danielNg91) [inline Api/Controllers/ActivityController.cs:27]: more specific name for this method. Don't use just Get, Post for controllers  methods

File: Api/Program.cs [hunk 21]
```diff
@@ -1,15 +1,25 @@
...
+using Api.Configurations;
 
-// Add services to the container.
+var builder = WebApplication.CreateBuilder(args);
 
 builder.Services.AddControllers();
-// Learn more about configuring Swagger/OpenAPI at https://aka.ms/aspnetcore/swashbuckle
 builder.Services.AddEndpointsApiExplorer();
 builder.Services.AddSwaggerGen();
+builder.Services.AddDbContext<ApplicationDbContext>
```
Comment (danielNg91) [inline Api/Program.cs:13]: map to configuration

File: Application/Interfaces/Activity/IActivityServices.cs [hunk 22]
```diff
@@ -0,0 +1,16 @@
+﻿using Domain.Entities.Activity;
+
+namespace Application.Interfaces.Activity;
+
+public interface IActivityServices
+{
+    //Activity
+    Task<ICollection<ActivityEntity>> List();
+    Task<ActivityEntity> GetById(Guid activityId); // Get by Activity Id
```
Comment (danielNg91) [inline Application/Interfaces/Activity/IActivityServices.cs:9]: what're these comments for?

File: Application/Interfaces/Activity/IActivityServices.cs [hunk 23]
```diff
@@ -0,0 +1,16 @@
...
+    Task<ActivityEntity> GetByName(string Name); // Get by Activity Name
+    Task<ICollection<ActivityEntity>> GetByTypeId(Guid activityTypeId); // Get by Activity Type Id
```
Comment (danielNg91) [inline Application/Interfaces/Activity/IActivityServices.cs:11]: how about using overload for 2 methods?

File: Application/Services/ActivityServices/ActivityTypeServices.cs [hunk 24]
```diff
@@ -0,0 +1,75 @@
...
+        var activityTypes = await _activityTypeRepo.ListAsync();
+        return activityTypes;
+    }
+    public async Task<ActivityType> GetById(Guid activityTypeId)
+    {
+        // Get by Activity Type Id
+        var activityType = await _activityTypeRepo.FindByIdAsync(activityTypeId);
+        if (activityType == null)
+        {
+            throw new Exception($"Activity type not found");
```
Comment (danielNg91) [inline Application/Services/ActivityServices/ActivityTypeServices.cs:28]: fixed by yesterday's comment

File: Application/Services/AssetServices/AssetAttributeServices.cs [hunk 25]
```diff
@@ -0,0 +1,72 @@
...
+        {
+            throw new Exception($"Asset attribute not exist");
+        }
+        else
+        {
+            return assAtt;
+        }
+    }
+    public async Task<ICollection<AssetAttribute>> GetById(Guid id, int typeId)
+    { // typeId: 1: AssetId, 2: AttributeGroupId
```
Comment (danielNg91) [inline Application/Services/AssetServices/AssetAttributeServices.cs:34]: why still use typeId??

**Review by danielNg91 — CHANGES_REQUESTED** [review]

File: Api/Configurations/AppConfigurationService.cs [hunk 26]
```diff
@@ -0,0 +1,52 @@
...
+            services.AddScoped<ILevelServices, LevelServices>();
+            services.AddScoped<IPaymentServices, PaymentServices>();
+            services.AddScoped<ITransactionServices, TransactionServices>();
+            services.AddScoped<IUserServices, UserServices>();
+            services.AddScoped<IWalletCategoryServices, WalletCategoryServices>();
+            services.AddScoped<IWalletServices, WalletServices>();
+        }
+        public static void AddDbServices(this IServiceCollection services)
+        {
+            services.AddDbContext<ApplicationDbContext>
```
Comment (danielNg91) [inline Api/Configurations/AppConfigurationService.cs:48]: map from appsettings.json to C# class

File: Application/Services/AttributeServices/AttributeServices.cs [hunk 27]
```diff
@@ -0,0 +1,55 @@
...
+        var attList = List();
+        var check = false;
+        foreach (var attgrp in attList.Result)
+        {
+            if (attgrp.Name == attributeGroup.Name)
+            {
+                check = true;
+                break;
+            }
+            else continue;
```
Comment (danielNg91) [inline Application/Services/AttributeServices/AttributeServices.cs:36]: this is end of for block, what is 'continue' use for?

File: Application/Services/AttributeServices/AttributeServices.cs [hunk 28]
```diff
@@ -0,0 +1,55 @@
...
+        }
+        var attGrp = new AttributeGroup{
+            Name = attributeGroup.Name,
+            Effect = attributeGroup.Effect
+        };
+        if (check == false)
```
Comment (danielNg91) [inline Application/Services/AttributeServices/AttributeServices.cs:42]: what you will do when check  == true?

File: Application/Services/WalletServices/WalletCategoryServices.cs [hunk 29]
```diff
@@ -0,0 +1,42 @@
...
+    {
+        return await _walletCategoryRepo.FindByIdAsync(categoryId);
+    }
+    public async Task<ICollection<WalletCategory>> GetByGameId(Guid gameId)
+    {
+        var walCat = await _walletCategoryRepo.WhereAsync(
+            wc => wc.GameId.Equals(gameId));
+        if (walCat.Count == 0)
+        {
+            throw new Exception($"Wallet category or game not found");
```
Comment (danielNg91) [inline Application/Services/WalletServices/WalletCategoryServices.cs:29]: why you still throw an Exception. Should remember, Exception is the same as reponse with 500 status.

**Review by danielNg91 — CHANGES_REQUESTED** [review]

**Pull request closed without merging**